### ISSUE TITLE
[Feature Request] Multimodal MoE support: MiniMax-M3 (vision-language) + activate existing VL weights (Qwen3.5-MoE)

### ISSUE THREAD

**Issue by drunkcoding** [EfficientMoE/MoE-Infinity#221]
## Prerequisites

- [x] I have searched existing issues and reviewed documentation.

## Problem Description

MoE-Infinity currently serves **text-only** inputs. The serving path
(`moe_infinity/serving/model_runner.py`) prepares padded `input_ids` /
`attention_mask` / `position_ids` exclusively — there is no path for
`pixel_values` or any multimodal processor output. This blocks two things:

1. **MiniMax-M3** (`MiniMaxAI/MiniMax-M3`, `MiniMaxM3SparseForConditionalGeneration`)
   — a flagship multimodal MoE that is *exactly* the workload MoE-Infinity is
   built for: **~428B total / ~23B active**, **128 routed experts + 1 shared,
   top-4 routing**, 1M-token context, CLIP-style ViT vision encoder
   (336×336–2016×2016), MiniMax Sparse Attention (block-sparse), BF16/MXFP8
   weights. At 428B, no consumer/prosumer setup can hold it resident —
   expert offloading is the only realistic single-server path, yet we cannot
   run it because the input pipeline and model registry are text-only.
2. **Qwen3.5-MoE vision weights are already downloaded but dead.** We ship
   `Qwen3_5MoeForConditionalGeneration` support with vision and MTP weights
   present but explicitly unused (per README and
   `docs/model-compatibility.md`). Users pay the disk/offload cost of VL
   checkpoints and get text-only serving.

The core engine does **not** have this limitation: the offload/prefetch engine
(`moe_infinity/runtime/model_offload.py`) and expert tracer
(`moe_infinity/memory/expert_tracer.py`) operate purely on (layer, expert-id)
activation patterns and expert weight residency — they are modality-agnostic.
The gap is confined to input preparation, model wrappers, and serving-layer
plumbing. This feature closes that gap.

Note: this issue deliberately targets **multimodal understanding on MoE
backbones** (image/video in → text out). Omni-modal *generation* models such
as MiniMax-H3 (dense 33B diffusion-style AV generator, no routed experts) are
out of scope here and tracked separately in the companion RFC
(new sibling project proposal).

## Proposed Solution

Introduce a general **multimodal input path** plus per-model wrappers,
in stages:

### Stage 1 — Multimodal input plumbing (model-agnostic)

- Extend `ModelRunner` / batch structs (`moe_infinity/serving/`) to carry an
  optional `multimodal_inputs` dict (`pixel_values`, `image_grid_thw`,
  video tensors, etc.) alongside `input_ids`.
- Use `AutoProcessor` (falling back to `AutoTokenizer`) in entrypoints;
  OpenAI server accepts `image_url` / base64 content parts on
  `/v1/chat/completions` (OpenAI vision API shape).
- Vision encoder runs GPU-resident in the prefill step; its output embeddings
  are scatter-merged into the text embedding sequence (standard HF VL merge),
  after which decode is unchanged — expert offloading, paged KV, continuous
  batching all apply as-is.
- Scheduler/KV accounting: image token expansion must be counted at admission
  time (a single image can expand to thousands of tokens at high resolution).

**QA scenarios (Stage 1):**

1. Request parsing (CI, no GPU/model download):
   `pytest tests/python/serving/test_multimodal_inputs.py` (new) —
   (a) `POST /v1/chat/completions` with content parts
   `[{type:"text"},{type:"image_url", base64 64×64 PNG}]` against a
   registered **text-only** model → expect HTTP 400 with error message
   `"model does not support image input"`;
   (b) same request against a VL-registered mock model (stub processor) →
   request admitted; assert the built batch carries `pixel_values` with the
   processor-reported shape and that image-token expansion is reflected in
   the scheduler's admitted token count.
2. Text-only regression: `pytest tests/python/unit tests/python/serving` →
   pass set identical to the pre-change baseline (record baseline commit
   hash in the PR); byte-identical responses for a fixed seed text request
   replayed against the server before/after.

### Stage 2 — Qwen3.5-MoE vision activation (proof of pipeline)

- Smallest step with existing registered model: activate the already-present
  vision tower of `Qwen3_5MoeForConditionalGeneration`, keep the vision
  encoder resident (it is small relative to the 256 routed experts we already
  offload). This validates Stage 1 end-to-end with weights we already handle.

**QA scenarios (Stage 2):**

1. Image inference smoke (GPU host, ≥16GB VRAM + SSD):
   `CUDA_VISIBLE_DEVICES=0 python examples/vl_example.py --checkpoint
   Qwen/Qwen3.5-35B-A3B --image tests/fixtures/images/cat.jpg --prompt
   "What animal is in the picture?" --offload_dir /tmp/qwen35-vl`
   (new example script) → exit 0; generated answer contains `"cat"`
   (asserted in-script); expert-cache telemetry shows >0 routed-expert
   fetches (proves the offload path served a VL request).
2. Parity vs HF reference: one-time reference generation with plain
   `transformers` (`Qwen3_5MoeForConditionalGeneration` + `AutoProcessor`,
   greedy, `max_new_tokens=32`, fixed image+prompt) on a large-memory host;
   token ids stored as a fixture.
   `pytest tests/python/integration/test_qwen35_vl_parity.py` runs the same
   input through the `MoE` wrapper (greedy) → generated token ids exactly
   equal to the fixture.

### Stage 3 — MiniMax-M3 model support

- New wrapper under `moe_infinity/models/minimax_m3/` following the
  GLM-5.2 / DeepSeek-V4 pattern (conditional registration: only when
  `transformers` provides `MiniMaxM3SparseForConditionalGeneration`).
- Register in `moe_infinity/common/constants.py` + dispatch in
  `moe_infinity/entrypoints/big_modeling.py`.
- Expert offloading: 128 routed experts per MoE layer offloaded; shared
  expert, dense early layers, attention, vision tower stay resident
  (mirrors the GLM-5.2 resident/offloaded split).
- Attention: MiniMax Sparse Attention (block-sparse) — initial support via
  eager/PyTorch fallback (same approach we took for GLM-5.2 DSA:
  `attn_implementation="eager"`), optimized kernels as follow-up.
- Precision: BF16 first; MXFP8 routed-expert path as follow-up (we already
  have FP8-in-host-store precedent from GLM-5.2 and FP4 from
  DeepSeek-V4-Flash).

**QA scenarios (Stage 3):**

1. Mock parity (CI, no checkpoint download): `pytest
   tests/python/unit/test_minimax_m3_wrapper.py` (new) — build a tiny
   `MiniMaxM3` config (2 layers, 8 experts, top-2, tiny vision tower),
   random weights; run one forward through the MoE wrapper with experts
   offloaded to a tmpdir vs plain `transformers` forward → logits
   `allclose(rtol=1e-3)`, and offload-engine counters confirm expert
   fetches occurred (wrapper actually exercised the dispatch path).
   Mirrors the existing tiny-fixture pattern
   (`benchmarks/dflash/validate_unified_execution.py --fixture tiny`).
2. Real-checkpoint smoke (multi-GPU server with ≥1TB host RAM + NVMe;
   428B checkpoint):
   `python examples/vl_example.py --checkpoint MiniMaxAI/MiniMax-M3
   --image tests/fixtures/images/cat.jpg --prompt "What animal is in the
   picture?" --offload_dir /nvme/m3-store --device_memory_ratio 0.8`
   → exit 0; answer contains `"cat"`; script asserts
   `torch.cuda.max_memory_allocated()` stays under the configured budget
   on every visible GPU; expert cache hit-rate telemetry non-empty.
3. Small-eval sanity (same host): 20 image-question pairs from a fixed
   fixture list; assert ≥16/20 answers contain the expected keyword
   (threshold recorded in the test; guards against silently broken vision
   token merge while tolerating sampling-free generation variance across
   kernels).

- MXFP8 expert host store for M3.
- Block-sparse attention kernel integration (FlashInfer/custom).
- Video input support; multi-image; prefix caching interaction with image
  tokens.

### Acceptance criteria

- [ ] `MoE("MiniMaxAI/MiniMax-M3", {...})` loads with experts offloaded and
      answers an image+text prompt correctly (Stage 3 QA: mock logits
      parity in CI + real-checkpoint smoke + 20-pair keyword eval; full
      HF-reference parity at 428B is impractical and is covered at tiny
      scale by the mock test).
- [ ] OpenAI server accepts `image_url` content parts for registered VL
      models; text-only behavior is byte-identical for existing models.
- [ ] Qwen3.5-MoE serves image+text via the same path.
- [ ] Expert tracing/prefetch statistics populated identically for
      multimodal requests (engine modality-agnosticism preserved).
- [ ] `docs/model-compatibility.md` updated with the multimodal matrix.

## Alternatives Considered

1. **Serve M3 text-only first** (like Qwen3.5 today): lowest effort, but
   perpetuates the "VL weights are dead weight" problem and delivers little —
   M3's differentiation is multimodal + 1M context.
2. **Defer to vLLM/SGLang for multimodal MoE**: they support M3 but assume
   the model fits in aggregate GPU memory (multi-GPU EP/TEP). MoE-Infinity's
   niche — 428B-scale MoE on memory-constrained single servers via
   SSD/host-memory expert offload — is not covered by them.
3. **New repo for all multimodal models**: rejected for *understanding*
   models — the expert-offload engine already generalizes, and forking would
   duplicate the entire runtime/serving stack. A new repo **is** the right
   call for dense omni-modal *generation* (MiniMax-H3), tracked in the
   companion RFC.

## Additional Context

- Engine audit (this repo): offload engine
  (`moe_infinity/runtime/model_offload.py`) and tracer
  (`moe_infinity/memory/expert_tracer.py`) have no modality coupling;
  text assumptions are isolated to `moe_infinity/serving/model_runner.py`,
  entrypoints, and the OpenAI server request schema.
- MiniMax-M3: https://huggingface.co/MiniMaxAI/MiniMax-M3
- vLLM M3 serving notes (EP, MXFP8, parsers):
  https://vllm.ai/blog/2026-06-12-minimax-m3-vllm
- TensorRT-LLM M3 deployment guide (block-sparse attention requirements):
  https://nvidia.github.io/TensorRT-LLM/latest/deployment-guide/deployment-guide-for-minimax-m3-on-trtllm.html
- Precedent in-repo for conditional registration + mixed-precision expert
  stores: GLM-5.2 (`docs/glm-5.2.md`), DeepSeek-V4-Flash
  (`moe_infinity/models/deepseek_v4/README.md`).

## Importance

Important

## Usage Statistics (Optional)

Maintainer-driven roadmap item.

**Comment by drunkcoding** [comment]
Companion RFC for the out-of-scope dense omni-modal generation track (MiniMax-H3 → proposed sibling repo `Omni-Infinity`): #222

**Comment by drunkcoding** [comment]
**Open-source verification (pre-implementation gate, passed):** `MiniMaxAI/MiniMax-M3` weights are downloadable from HF and `MiniMaxM3SparseForConditionalGeneration` is native in `transformers 5.12.0` (verified against a local install — no `trust_remote_code` needed), matching this issue's conditional-registration plan. **License note:** M3 ships under the custom *MiniMax Community License* (commercial-use restrictions) — docs/model-compatibility.md entry must state this; it does not block implementation. `Qwen/Qwen3.5-35B-A3B` is Apache-2.0, fully open.

**Comment by drunkcoding** [comment]
**Location correction after the moe-store split landed (#214/#216):** model wrappers and the registry now live in `EfficientMoE/moe-store` (`moe_store/wrappers/`, `moe_store/registry/`), not `moe_infinity/models/` (deleted on main). Therefore:
- **Stage 1** (multimodal input plumbing: serving batch structs, AutoProcessor, OpenAI `image_url` parts, scheduler token accounting) stays in **MoE-Infinity** (`moe_infinity/serving/`, `entrypoints/`).
- **Stage 3** (MiniMax-M3 wrapper + registration) lands in **moe-store** (`moe_store/wrappers/minimax_m3.py` + registry), with engine-side glue here.
- Related in-flight vision work exists: branch `feat/dsv4-flash-vision-exp` and `moe_store/wrappers/deepseek_v4/VISION_EXP_NOTES.md` — Stage 1 design should align with it. Stage 1 implementation has started on branch `feat/multimodal-stage1`.

**Comment by drunkcoding** [comment]
**Stage 1 implemented** on branch `feat/multimodal-stage1` (3 commits ahead of main, not yet pushed):
- `b4ca32a` serving structs carry `multimodal_inputs` through sequence/batch/engine
- `d068953` model runner forwards multimodal model inputs
- `f70473f` OpenAI server accepts vision chat content parts (image_url/base64), AutoProcessor→AutoTokenizer fallback, clean 400 for non-VL models + Stage 1 QA tests

Verification: `tests/python/serving/test_multimodal_inputs.py` 4/4 passed (mock-based, no GPU); pre-commit (ruff/mypy) green; CPU regression suite shows no new failures vs recorded baseline (3 pre-existing failures unchanged). Next: PR + Stage 2 (Qwen3.5-VL activation).

**Comment by drunkcoding** [comment]
**Stage 3 (MiniMax-M3) merged to main** in #230.

- moe-store side (EfficientMoE/moe-store#8): `SyncMiniMaxM3VLSparseMoeBlock`, `minimaxm3` registry entry, expert-id parsing — released in **moe-store v0.2.1** (git tag; note the PyPI trusted-publisher config is broken — pre-existing, v0.2.0 failed identically — so PR CI installs from the tag).
- Engine side: 128 routed experts offloaded; text backbone, shared expert, vision tower, and multimodal projector resident; HF eager attention path (block-sparse Lightning Indexer kernels are a follow-up).
- CI-safe coverage: sync-vs-HF block parity (max abs diff 7.5e-08), residency policy, registry, topology.

Remaining Stage 3 follow-ups from this issue: MXFP8 routed-expert host store, block-sparse attention kernels, real-checkpoint smoke + 20-pair keyword eval (needs the ≥1TB RAM + NVMe rig).
